### DISCUSSION TITLE
Echo Cancellation Solutions for Continuous Microphone Input in Deepgram TTS/STT Streaming

### DISCUSSION THREAD

**Discussion by lincolnsmiff** [deepgram/community#866] [General help] [answered]
We are using Deepgram TTS and STT streaming in our application. Our current setup involves turning the microphone on to capture user input, then off to process the reply, and back on again for more input. This approach avoids echo cancellation issues in software where our bot ends up conversing with itself. However, it prevents users from interrupting the bot mid-conversation.

We are implementing an audio hardware solution for echo cancellation, but we seek a software-level solution as well. Is there a feature in the API, such as speaker diarization (speaker1, speaker2, but ignore speaker2) or any other method, that can help with echo cancellation and allow us to keep the microphone input on continuously for a better user experience?

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by nikolawhallon** [comment, marked as answer]
There is no feature that we provide that can help with this, but it's definitely something I've encountered frequently! The Web APIs offer echo cancellation features which make web-based clients function pretty well for your use-case (things like https://developer.mozilla.org/en-US/docs/Web/API/MediaTrackSettings/echoCancellation) - I imagine there must be some echo cancellation libraries for desktop applications/CLIs too (I'm finding some examples in Python here and there), and luckily this isn't so much an issue when calling a bot via Twilio's APIs as phones do echo cancellation as well.
